Fix internal assert custom function when input does not require grad

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72133
* __->__ #72008

Fixes  #71119

Technically BC-breaking because when an input does not require grad, previously it was returned as-is instead of a view because it didn't need to. Now we will also return a view in that case (whether or not forward AD runs).

Differential Revision: [D33859553](https://our.internmc.facebook.com/intern/diff/D33859553)

soulitzer committed:
Fix internal assert custom function when input does not require grad

[ghstack-poisoned]

soulitzer committed:
Update on "Fix internal assert custom function when input does not require grad"




[ghstack-poisoned]

soulitzer committed:
Update on "Fix internal assert custom function when input does not require grad"




[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33859553).

soulitzer committed:
Update on "Fix internal assert custom function when input does not require grad"


Fixes  #71119

Technically BC-breaking because when an input does not require grad, previously it was returned as-is instead of a view because it didn't need to. Now we will also return a view in that case (whether or not forward AD runs).

Differential Revision: [D33859553](https://our.internmc.facebook.com/intern/diff/D33859553)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33859553).

albanD commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -215,6 +215,13 @@ optional_variable_list _process_backward_mode_ad(
+          // We need this in case forward AD runs, since forward AD assumes that the output
+          // processed by backward ad can never be an input as-is.
+          // If forward AD does not run, we do some extra work here, but that is OK
+          AutoGradMode grad_mode(false);
```
- Can we have this in a separate function that is called here and below?
- Didn't we add a context to disable forward mode AD while doing this view in a previous PR??

soulitzer replied:
- sounds good
- that was the initial plan in the previous PR, but we eventually decided not to in favor of just returning early in the _process_forward_ad function. (we did disable forward AD while calling user-defined jvp)

soulitzer replied:
Hmm maybe I am missing something? My understanding is that we had two options:
1. if forward is no-op, we require user to manually return a view in jvp (otherwise error) to match the behavior of what backward does (when the output does require grad)
2. if forward is no-op, automatically return a view in jvp (we did this by continuing to run the view with forward grad enabled, and just adding a `continue` as we loop through the outputs in _process_forward_ad, as to avoid setting the forward grad again).

We ended up going with (2), but we didn't consider the case in _process_backward_ad when the output is considered non-differentiable (perhaps because it does not require grad), so in this PR, we add performing the view in this case as well.

soulitzer committed:
Update on "Fix internal assert custom function when input does not require grad"


Fixes  #71119

Technically BC-breaking because when an input does not require grad, previously it was returned as-is instead of a view because it didn't need to. Now we will also return a view in that case (whether or not forward AD runs).

Differential Revision: [D33859553](https://our.internmc.facebook.com/intern/diff/D33859553)

[ghstack-poisoned]

soulitzer committed:
Update on "Fix internal assert custom function when input does not require grad"


Fixes  #71119

Technically BC-breaking because when an input does not require grad, previously it was returned as-is instead of a view because it didn't need to. Now we will also return a view in that case (whether or not forward AD runs).

Differential Revision: [D33859553](https://our.internmc.facebook.com/intern/diff/D33859553)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -5588,7 +5588,9 @@ def jvp(ctx, nt, x_t, nt2, y_t):
-        for jvp_mul_by_2 in (True, False):
+        error_regex = "A custom Function's forward is returning a view \\(or an input as-is\\)"
+
+        for returns_view in (True, False):
```
Why not add a 3rd case instead of removing one of the existing ones?

albanD reviewed (approved):
LGTM!

soulitzer committed:
Update on "Fix internal assert custom function when input does not require grad"


Fixes  #71119

Technically BC-breaking because when an input does not require grad, previously it was returned as-is instead of a view because it didn't need to. Now we will also return a view in that case (whether or not forward AD runs).

Differential Revision: [D33859553](https://our.internmc.facebook.com/intern/diff/D33859553)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33859553).

albanD reviewed (approved):
Thanks for the update!

anjali411 commented:
@soulitzer could you please add the release notes labels?
